connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java

```diff
@@ -434,50 +516,93 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
-        Class<? extends HeaderConverter> klass = null;
+        return getHeaderConverter(config, classPropertyName, null, classLoaderUsage);
+    }
+
+    public HeaderConverter newHeaderConverter(AbstractConfig config, String classPropertyName, String versionPropertyName) {
+        ClassLoaderUsage classLoader = config.getString(versionPropertyName) == null ? ClassLoaderUsage.CURRENT_CLASSLOADER: ClassLoaderUsage.PLUGINS;
+        return getHeaderConverter(config, classPropertyName, versionPropertyName, classLoader);
+    }
+
+    private HeaderConverter getHeaderConverter(AbstractConfig config, String classPropertyName, String versionPropertyName, ClassLoaderUsage classLoaderUsage) {
+        if (!config.originals().containsKey(classPropertyName)) {
+            // This configuration does not define the Header Converter via the specified property name
+            return null;
+        }
+
+        HeaderConverter plugin = getVersionedPlugin(config, classPropertyName, versionPropertyName,
+                HeaderConverter.class, classLoaderUsage, scanResult.headerConverters());
+
+        String configPrefix = classPropertyName + ".";
+        Map<String, Object> converterConfig = config.originalsWithPrefix(configPrefix);
+        converterConfig.put(ConverterConfig.TYPE_CONFIG, ConverterType.HEADER.getName());
+        log.debug("Configuring the header converter with configuration keys:{}{}", System.lineSeparator(), converterConfig.keySet());
+
+        try (LoaderSwap loaderSwap = withClassLoader(plugin.getClass().getClassLoader())) {
+            plugin.configure(converterConfig);
+        }
+        return plugin;
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private <U> U getVersionedPlugin(
+            AbstractConfig config,
+            String classPropertyName,
+            String versionPropertyName,
+            Class basePluginClass,
+            ClassLoaderUsage classLoaderUsage,
+            SortedSet<PluginDesc<U>> availablePlugins
+    ) {
+
+        String version = versionPropertyName == null ? null : config.getString(versionPropertyName);
+        VersionRange range = null;
+        if (version != null) {
+            try {
+                range = VersionRange.createFromVersionSpec(version);
+            } catch (InvalidVersionSpecificationException e) {
+                throw new ConnectException(String.format("Invalid version range for %s: %s %s", classPropertyName, version, e));
+            }
+        }
+
+        Class<? extends U> klass = null;
+        String basePluginClassName = basePluginClass.getSimpleName();
         switch (classLoaderUsage) {
             case CURRENT_CLASSLOADER:
-                if (!config.originals().containsKey(classPropertyName)) {
-                    // This connector configuration does not define the header converter via the specified property name
-                    return null;
-                }
```

Comment:
Odd, the newHeaderConverter didn't perform this containsKey check when PLUGINS was specified, when the configuration is coming from the worker config.

But since `header.converter` has a default, it was always valid to call getClass. Now the `getHeaderConverter` method performs this check unconditionally, does this mean that Worker#startTask will sometimes not instantiate a header converter?

Reply:
Good point, I think if the header converter is not defined in both connector config and worker config it will not be initialized.

TBH, this initialization mechanism always seemed a bit brittle. Class loader usage seems to be strongly tied to whether we use the connector config or the worker config. Although we don't do it, there is nothing stopping us from passing a WorkerConfig with current class loader or vice versa. Also, CURRENT_CLASSLOADER seems to indicate that we are creating the plugin with the loader of the current thread. This is not exactly the case as its actually the loader in use when the AbstractConfig was initialized (which internally does Utils.loadclass for class type configs). 
This can be addressed separately but for now I have preserved the existing behavior to reduce potential edge cases.

Reply:
> Although we don't do it, there is nothing stopping us from passing a WorkerConfig with current class loader or vice versa. 

Yeah we don't do that because it doesn't make much sense. CURRENT_CLASSLOADER implies that there's a particular PluginClassLoader which is given precedence. The WorkerConfig doesn't have a preferred PluginClassLoader, so it only ever makes sense to use PLUGINS.

Although CURRENT_CLASSLOADER when the TCCL is the DelegatingClassLoader should be equivalent to PLUGINS as far as I understand...

I agree that we don't have to fully clean-up the ClassLoaderUsage stuff immediately, just enough to keep compatible behaviors.